### DISCUSSION TITLE
Why must there be 1 language per file extension? Seems very restrictive.

### DISCUSSION THREAD

**Discussion by stephen-carter-at-sf** [pmd/pmd#5343] [General]
Hello,

I work on the Salesforce Code Analyzer product at Salesforce and we use PMD quite a bit. One fundamental flaw that I am finding, that our Security team has to work around, is the fact that we can't associate `*.meta.xml` files to our own SFMETA language because the XmlLanguageModule already registers the `xml` extension.

I see that it appears that TextFile has a LanguageVersion. This seems the root of the restriction if I understand this code correctly: 
https://github.com/pmd/pmd/blob/main/pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFile.java#L128

If a company, like Salesforce, has their own xml files that they want to be able to add rules to. But we don't want our rules running on all xml files - but only .meta.xml files for example... then there appears to be no clean way of doing this.

Also, we have legacy file extensions like `*. quickAction`, `*. flow`, `*.flexipage` etc which are all also xml based. So if we write a rule that should work for both *.quickAction or *.meta.xml files - then we need to add in a new SFMETA language - but we can't add in *.meta.xml to it because https://github.com/pmd/pmd/blob/main/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageModuleBase.java#L285 only wants to take in the last part of the extension without the "." and code like https://github.com/pmd/pmd/blob/main/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageFilenameFilter.java#L56 only gets the last part from the file name.

Because of this, we have to have duplicate rule definitions. One for SFMETA language and one for XML.

Ideally, I would recommend that multiple languages should be able to accept the same file extensions. Then you run the rules for a language based on the subset of files that it accepts. I would also recommend that we change the way the file filtering works so that it just checks if the file name ends with the provided extension so that ".meta.xml" can be given instead of "xml" in the extensions of the LanguageModuleBase. 

So either we could map 1 file extension to multiple languages (preferable)... or alternatively map 1 rule to multiple languages.

--
What are the PMD developer's thoughts on this?

**Comment by jsotuyod** [comment]
@stephen-carter-at-sf yes, this is something that has been in our backlog for quite some time (see #461). XML files are the most common problematic extension, but overall, language detection / extension mapping is something that comes up from time to time (ie: #4780).

To a certain degree, this has been improved on with the addition of the `-force-language` flag, which allows to ignore extension mapping rules and forcefully apply a language… but it requires a dedicated PMD run, so it's applicability is limited.

As for the preferred approach, there are 2 that have seemed the most promising so far:
- A single file can be matched to multiple languages
- We introduce a new concept, such as a "dialect" that is a specific extension of a given language (ie: a POM is a specific type of XML, as is XSLT)

The dialect approach is very interesting in:
- it explicitly states the lexer / parser should be the same one
- it allows to retain the single read / parse of each file
- it naturally allows to apply any rule from the root language or the dialect

On top of that we have the language detection limitations and how we ensure flexibility.
* if I want to create an XPath rule for a specific XML dialect, do I need to register it in code? or is it declarative in some way?
* how can I add / alter extensions used by a file (ie: twig for html), can I do it without the limitations of `-force-language`

There is a draft #4945 where a lot of these discussions have continued, but was halted without a definition. This is unfortunately something we are not currently working on.

Most probably than not, this will introduce breaking changes and therefore be deferred to PMD 8, but definitely something we need to start looking into if we expect it to make it in time.

**Reply by stephen-carter-at-sf** [reply]
OK I think for now, we will simply make multiple passes of PMD. We use the java api, and I see I can maybe separate out the files that are associated with xml language and run a separate pass of PMD using setForceLanguageVersion on the pmd configuration:

https://docs.pmd-code.org/apidocs/pmd-core/7.7.0/net/sourceforge/pmd/AbstractConfiguration.html#setForceLanguageVersion(net.sourceforge.pmd.lang.LanguageVersion)

**Reply by jsotuyod** [reply]
Yes, that should work for now.

Yesterday I spent some thought on this problem, I expect to work on a PoC for language dialects once I get some spare time.

**Reply by oowekyala** [reply]
@jsotuyod FWIW I had started working on a PoC in https://github.com/oowekyala/pmd/tree/core-lang-dialects

**Reply by jsotuyod** [reply]
@oowekyala thanks. My approach was very similar, except I was aiming to avoid the `extends AnotherLanguageModule`, as it forces the existence of a new constructor, and therefore languages need to "opt-in" into supporting dialects… Instead, I expected the `LanguageMetadata.isDialectOf` method to internally imply also a `dependsOnLanguage` to ensure runtime validation.

I was also thinking of changing the naming / id convention for dialects to make it explicit, such as "base-lang/dialect", ie: "xml/pom" and "xml/xslt" (which would be a breaking change and need to keep the standalone language for backwards compatibility until PMD 8).

Your code on `RuleSet` also needs some extra tailoring, as dialect versions and base language versions are different, so I had some thought into how to handle this. And probably some extra abstractions to help extend a Language Handler (ie: add new xpath functions, metrics, etc.) ideally without needing to have a compile time-dependency / extend; but you already covered roughly 90% of what's needed.

If I get around it, I'll still take your code as a starting point.

**Reply by jsotuyod** [reply]
@stephen-carter-at-sf please, feel free to check #5438 and provide feedback

**Reply by jsotuyod** [reply]
@stephen-carter-at-sf  this has now been merged and is part of today's release of PMD 7.13.0.
Looking forward to see what you build with this and any feedback you provide.
